### PR TITLE
Feature/mem opt pass refactor

### PR DESCRIPTION
**Pull request by sneaxiy**
This PR refactors legacy C++ memory optimization strategy, which is based on changing variable names. The new strategy uses `ShareTensorBufferOpHandle` to perform memory reuse during run-time instead of changing variable names.

This is designed to fix the issue that users have to set `var.persistable = True` before using this memory optimization strategy.

**Sorry that: since this PR is too long, I would remove the legacy C++ memory optimization codes in next PR.**

I have tried to turn on `build_strategy.memory_optimize = True` by default on CE Debug. All tasks have passed. Please see [here](http://ce.paddlepaddle.org:8080/viewLog.html?buildId=54043&buildTypeId=PaddleModesl_ModelDebug_Model11Build&tab=buildResultsDiv).

### CODE REVIEW

File: paddle/fluid/framework/details/share_tensor_buffer_functor.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#include <unordered_set>
+#include <utility>
+#include <vector>
+#include "paddle/fluid/framework/details/op_handle_base.h"
+
+namespace paddle {
+namespace framework {
+
+class Variable;
+class Scope;
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/share_tensor_buffer_functor.h]: Is there a way we don't have to declare here?

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+static NodeDependency ReverseNodeDependency(NodeDependency dep) {
+  return dep == NodeDependency::kBefore
+             ? NodeDependency::kAfter
+             : (dep == NodeDependency::kAfter ? NodeDependency::kBefore : dep);
+}
+
+class BufferSharedCrossOpMemoryReusePass : public MemoryReusePass {
+ protected:
+  std::string ReuseType() const override { return "cross_op_memory_reuse"; }
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc:49]: Should we use const value for this kind of string if it may be referred somewhere?

File: paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc [hunk 3]
```diff
@@ -131,14 +120,24 @@ void MemoryReusePass::CollectShareTensorBufferOpHandles() const {
...
+          reused_var_pair.first);
+      reused_out_var_names_[pair.first->GetScopeIdx()].insert(
+          reused_var_pair.second);
+    }
   }
 }
 
-bool MemoryReusePass::IsVarAlreadyReused(details::VarHandle *var) const {
-  return reused_var_names_[var->scope_idx()].count(var->Name()) > 0;
+bool MemoryReusePass::IsInVarAlreadyReused(details::VarHandle *in_var) const {
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc]: Maybe in future PR, because I saw a lot in this PR.

In general, there are some exceptions but input arguments are values or const references while output arguments are pointers.

Ref: https://google.github.io/styleguide/cppguide.html#Reference_Arguments

File: python/paddle/fluid/tests/unittests/parallel_executor_test_base.py [hunk 4]
```diff
@@ -65,8 +65,10 @@ def run_executor(exe, binary, feed, fetch_list):
                 main.random_seed = seed
 
             loss = method(use_feed=feed_dict is not None)
-            if memory_opt or use_ir_memory_optimize:
+            if memory_opt:
                 loss.persistable = True
+            else:
+                loss.persistable = False
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/parallel_executor_test_base.py]: loss.persistable = memory_opt

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc:49]: This string is only used in logging without referencing in other places. Maybe it is not necessary to use const value.

File: paddle/fluid/framework/details/share_tensor_buffer_functor.h (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/share_tensor_buffer_functor.h]: Done by including header files.

File: paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc]: Done.

File: python/paddle/fluid/tests/unittests/parallel_executor_test_base.py (hunk 4 above)
Comment (sneaxiy) [reply python/paddle/fluid/tests/unittests/parallel_executor_test_base.py]: Done.

File: paddle/fluid/framework/details/share_tensor_buffer_functor.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    const std::vector<ir::MemOptVarInfo *> &in_var_infos,
+    const std::vector<std::string> &out_var_names)
+    : scope_(scope),
+      scope_idx_(scope_idx),
+      op_type_(op_type),
+      in_var_infos_(in_var_infos),
+      out_var_names_(out_var_names) {
+  PADDLE_ENFORCE_EQ(in_var_infos_.size(), out_var_names_.size());
+  for (size_t i = 0; i < in_var_infos_.size(); ++i) {
+    Add(in_var_infos_[i], out_var_names_[i]);
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/share_tensor_buffer_functor.cc]: The function name, `Add`, is confusing.

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+    if (iter->second != 0) {
+      ++iter;
+      continue;
+    }
+
+    visited_ops.insert(iter->first);
+    ready_ops.push(iter->first);
+    sorted_ops.emplace_back(iter->first);
+    op_deps.erase(iter++);
+  }
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc]: I think that line 122~132 can be placed to a function of `OpGraphView`, because those code are used many times.

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+
+ private:
+  mutable Graph *graph_;
+
+  // All ops in the graph, grouped by scope index
+  mutable std::vector<std::vector<ComputationOpHandle *>> ops_;
+
+  // Index of each op in `ops_`, grouped by scope index.
+  // Index of each op is the index inside `deps_`.
+  mutable std::vector<std::unordered_map<ComputationOpHandle *, size_t>>
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc]: It is better that using `const ComputationOpHandle *` here.

File: paddle/fluid/framework/details/share_tensor_buffer_functor.cc (hunk 5 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/share_tensor_buffer_functor.cc]: Done. `Add` -> `AddReuseVarPair`.

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc (hunk 6 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc]: Done. Add `BreadthFirstVisit` method in `OpGraphView` and simplify codes.

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc (hunk 7 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc]: Done.

File: paddle/fluid/framework/ir/pass.cc [hunk 8]
```diff
@@ -37,7 +37,8 @@ Graph* Pass::Apply(Graph* graph) const {
   ApplyImpl(graph);
   // TODO(panyx0718): Add more verifications.
   PADDLE_ENFORCE(!HasCircle(*graph),
-                 "Illegal Pass. Generated graph shouldn't has cycle.");
+                 "Illegal Pass %s. Generated graph shouldn't has cycle.",
+                 Type());
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/pass.cc:41]: Grammar: shouldn't have

**Review by zhhsplendid — DISMISSED** [review]

File: paddle/fluid/framework/ir/pass.cc (hunk 8 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/pass.cc:41]: Done.

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+
+  for (auto *op : ops) {
+    // Transverse all output args of op, find whether there is unlived var
+    // can be reused.
+    auto out_args = op->Node()->Op()->OutputArgumentNames();
+    for (auto &out_arg : out_args) {
+      auto out_nodes = this->FindNodesByName(out_arg, op->Node()->outputs);
+      if (out_nodes.size() != 1) {
+        continue;
+      }
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc:179]: Why the out_nodes may be empty?

File: paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc [hunk 10]
```diff
@@ -171,57 +174,88 @@ MemoryReusePass::InsertShareTensorBufferOpHandleToGraph(
...
   }
 
-  if (in_var_desc->GetType() != proto::VarType::LOD_TENSOR ||
-      out_var_desc->GetType() != proto::VarType::LOD_TENSOR) {
+  if (out_var_desc->GetType() != proto::VarType::LOD_TENSOR) {
     return false;
   }
 
-  if (!FindNodesByName(in_name, out_var->GeneratedOp()->Node()->outputs)
+  if (!FindNodesByName(out_name, out_var.GeneratedOp()->Node()->inputs)
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc:254]: Please add what this check for.

File: paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc (hunk 9 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc:179]: It may happen when `out_arg == kEmptyVarName`.

**Review by chengduoZH — APPROVED** [review]

File: paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc (hunk 10 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.cc:254]: I have add corresponding comments before the method. Thanks!
